### DISCUSSION TITLE
Set cell temperature as input (Isothermal model)

### DISCUSSION THREAD

**Discussion by meckwell** [pybamm-team/PyBaMM#2527] [Q&A] [answered]
Dear all,

I wanted to set cell temperature(that varies with time) as input in a constant current discharge simulation. So I chose the isothermal model and tried to set the cell temperature through _**Ambient temperature**_ and load the external temperature data as input. But I encountered an error as follows: 

![Error](https://user-images.githubusercontent.com/33289036/205851217-ca69ec8c-30d3-4dda-9d4a-9c1cd6097c60.PNG)

I was wondering whether set cell temperature as input is doable in PyBaMM. Many thanks in advance.

My implementation is attached as follows for your reference:

# import temperatures  from file
os.chdir('/0Work/OneDrive - Imperial College London/0Postdoc at Imperial College/2021.11- First model') 
Dirc_ECN = "./ECN MODEL/Kokam7point5Ah/Simulation results/"# the folder you store results for a specific condition
temp_fun = pd.read_csv(Dirc_ECN+"1C_CC_48LP_CAN=Yes_CColl=Yes_T=0.0.csv", encoding_errors='ignore', header=1).to_numpy()
                       
# load parameter values
os.chdir('/0Work/OneDrive - Imperial College London/0Postdoc at Imperial College/2021.11- First model/PyBaMM/Kokam 7.5 Ah')
param = model.default_parameter_values

# create interpolant - must be a function of *dimensional* time
timescale = param.evaluate(model.timescale)
temp_interpolant = pybamm.Interpolant(temp_fun[:, 0], temp_fun[:, 2], timescale * pybamm.t)

# set temperatures  
param["Ambient temperature [K]"] = temp_interpolant
param["Initial temperature [K]"] = 278.15
capacity = param["Nominal cell capacity [A.h]"]
param.update({"Current function [A]": 1*capacity})

# set up simulation - for drive cycles we recommend using the CasadiSolver in "fast" mode
solver = pybamm.CasadiSolver(mode="fast")
simulation = pybamm.Simulation(model, parameter_values=param, solver=solver)

# simulate 
simulation.solve([0, 3500])

**Comment by DrSOKane** [comment]
Have you tried this?

param["Ambient temperature [K]"] = pybamm.InputParameter("T")
simulation = pybamm.Simulation(model=model, parameter_values=param, solver=solver)
simulation.solve([0, 3500], inputs={"T": temp_interpolant})

If that doesn't work, try setting the CasadiSolver to safe mode to see if you're hitting any events.

**Reply by meckwell** [reply]
Thank you Simon for the promptly response.

I just tried to use the lines you recommended. It hit the event as follows:
![image](https://user-images.githubusercontent.com/33289036/205913720-49703fcf-655f-4252-a92b-870a372760b7.png)
 
And I also tried Safe mode, it hit the same event as before
SolverError: Could not find consistent states: Could not find acceptable solution: solver returned NaNs

**Reply by rtimms** [reply]
The solver is struggling to find consistent initial conditions. Does your code work if the temperature is just a scalar? Could test with the first value in your temperature data.

**Reply by DrSOKane** [reply]
Hi Han, setting a function as an InputParameter doesn't work. You have to define the function then use

param.update({"Ambient temperature [K]": temp_interpolant})

and it should run. Sorry for the delay, that confused me as well!

**Reply by meckwell** [reply]
Dear Simon and Robert，

Apologies for not getting back to you sooner. Many thanks for your help! Robert's example code works!

Best regards,

Han

**Comment by rtimms** [comment, marked as answer]
Try 
```pythion3
import pybamm
import numpy as np

model = pybamm.lithium_ion.DFN()
param = pybamm.ParameterValues("Marquis2019")
timescale = param.evaluate(model.timescale)

# make dummy temperature data
t = np.linspace(0, 3600, 360)
temp = 298 + np.sin(2 * np.pi * t)
temp_interpolant = pybamm.Interpolant(t, temp, timescale * pybamm.t)

param["Ambient temperature [K]"] = temp_interpolant
param["Initial temperature [K]"] = 298

sim = pybamm.Simulation(model, parameter_values=param)
sim.solve([0, 3600])
sim.plot(
    ["Current [A]", "Terminal voltage [V]", "Volume-averaged cell temperature [K]"]
)
```
